src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,128 @@
+
+If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be "initialized" with no overhead and then written to in parts.
+
+## Sources of uninitialized memory
+
+### `mem::uninitialized()` and `MaybeUninit::assume_init()`
+
+[`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's basically _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB.
+
+Use [`MaybeUninit<T>`] instead.
+
+It is still possible to create uninitialized memory using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.
+
+### Padding
+
+Padding bytes in structs and enums are often but not always uninitialized. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are ginning up uninitialized `u8`s.
+
+Reading from padding [always produces uninitialized values][pad-glossary].
+
+
+
+### Moved-from values
+
+The following code is UB:
+
+```rust
+# use std::ptr;
+let x = String::new(); // String is not Copy
+let mut v = vec![];
+let ptr = &x as *const String;
+
+v.push(x); // move x into the vector
+
+unsafe {
+    // reads from moved-from memory
+    let ghost = ptr::read(ptr);
+}
+```
+
+Any type of move will do this, even when you "move" the value into a different variable with stuff like `let y = x;`.
+
+Note that Rust does let you "partially move" out of fields of a struct, in such a case the whole struct is now no longer a valid value for its type, but you are still allowed to "use" the struct to look at other fields. When doing such things, make sure there are no pointers that still think the struct is whole and valid.
+
+#### Caveat: `ptr::drop_in_place()`, `ManuallyDrop::drop()`, and `ptr::read()`
+
+[`ptr::drop_in_place()`] and [`ManuallyDrop::drop()`] are interesting: they call all the destructor[^1] on a value (or a pointed-to value in the case of the former). From a safety point of view they are identical; they are just different APIs for dealing with manually calling drop glue.
+
+[`ManuallyDrop::drop()`] makes the following claim:
+
+> Other than changes made by the destructor itself, the memory is left unchanged, and so as far as the compiler is concerned still holds a bit-pattern which is valid for the type T.
+
+In other words, Rust does _not_ consider these operations to do the same invalidation as a regular "move from" operation, even though they have a similar feel.
```

Comment:
I'm getting tripped up by some terminology here. I'm pretty sure that the kind of "move from" operation mentioned here doesn't "leave anything behind" - not even memory to access. In a sense, the builtin "move from" operation is either:

- Renaming an existing value
- or performing a typed copy out of some stack-allocated value (the stack may no longer contain the memory for the value afterward)

Reply:
Hmm, how would you frame this?

Reply:
As I said earlier I don't think this is super different from sticking an uninitialized value in the original memory

Reply:
I've always thought of `ptr::read` as the "move from" operation, with the additional freeing as a separate step that occurs right afterward. Maybe it could be viewed as "as soon as a memory location is no longer used, it's undef".

Reply:
Hmm, that makes sense. I'm heading up the hill to do some trailwork in a bit, but if you'd like feel free to merge this and propose alternate wording? Or we can leave this open and try and figure it out together.

Reply:
I've reworked this a bit, thoughts?